File: src/lib/api/fact.ts
import { invoke } from '@tauri-apps/api/core';

import type { Fact, FactCharacter } from './types';

export const factApi = {
	create: (request: {
		content: string;
		category?: string;
		revealed_in_scene_id?: string;
		status?: string;
	}) => invoke<Fact>('create_fact', { request }),

	getAll: () => invoke<Fact[]>('get_facts'),

	get: (id: string) => invoke<Fact>('get_fact', { id }),

	update: (
		id: string,
		request: {
			content?: string;
			category?: string;
			revealed_in_scene_id?: string;
			status?: string;
		}
	) => invoke<Fact>('update_fact', { id, request }),

	delete: (id: string) => invoke<void>('delete_fact', { id }),

	getForScene: (sceneId: string) => invoke<Fact[]>('get_facts_for_scene', { sceneId }),

	linkCharacter: (factId: string, bibleEntryId: string, learnedInSceneId?: string) =>
		invoke<FactCharacter>('link_character_to_fact', {
			factId,
			bibleEntryId,
			learnedInSceneId: learnedInSceneId ?? null,
		}),

	unlinkCharacter: (factId: string, bibleEntryId: string) =>
		invoke<void>('unlink_character_from_fact', { factId, bibleEntryId }),

	getCharacters: (factId: string) => invoke<FactCharacter[]>('get_fact_characters', { factId }),

	getCharacterKnowledge: (bibleEntryId: string, sceneId: string) =>
		invoke<Fact[]>('get_character_knowledge_at_scene', { bibleEntryId, sceneId }),
};

File: src-tauri/src/commands/fact.rs
use crate::models::{CreateFactRequest, Fact, FactCharacter, UpdateFactRequest};
use crate::AppState;
use tauri::State;

#[tauri::command]
pub fn create_fact(state: State<'_, AppState>, request: CreateFactRequest) -> Result<Fact, String> {
    let db = state.get_db()?;
    let db = db.as_ref().ok_or("No project open")?;
    db.create_fact(&request)
}

#[tauri::command]
pub fn get_facts(state: State<'_, AppState>) -> Result<Vec<Fact>, String> {
    let db = state.get_db()?;
    let db = db.as_ref().ok_or("No project open")?;
    db.get_facts()
}

#[tauri::command]
pub fn get_fact(state: State<'_, AppState>, id: String) -> Result<Fact, String> {
    let db = state.get_db()?;
    let db = db.as_ref().ok_or("No project open")?;
    db.get_fact(&id)
}

#[tauri::command]
pub fn update_fact(
    state: State<'_, AppState>,
    id: String,
    request: UpdateFactRequest,
) -> Result<Fact, String> {
    let db = state.get_db()?;
    let db = db.as_ref().ok_or("No project open")?;
    db.update_fact(&id, &request)
}

#[tauri::command]
pub fn delete_fact(state: State<'_, AppState>, id: String) -> Result<(), String> {
    let db = state.get_db()?;
    let db = db.as_ref().ok_or("No project open")?;
    db.delete_fact(&id)
}

#[tauri::command]
pub fn get_facts_for_scene(
    state: State<'_, AppState>,
    scene_id: String,
) -> Result<Vec<Fact>, String> {
    let db = state.get_db()?;
    let db = db.as_ref().ok_or("No project open")?;
    db.get_facts_for_scene(&scene_id)
}

#[tauri::command]
pub fn link_character_to_fact(
    state: State<'_, AppState>,
    fact_id: String,
    bible_entry_id: String,
    learned_in_scene_id: Option<String>,
) -> Result<FactCharacter, String> {
    let db = state.get_db()?;
    let db = db.as_ref().ok_or("No project open")?;
    db.link_character_to_fact(&fact_id, &bible_entry_id, learned_in_scene_id.as_deref())
}

#[tauri::command]
pub fn unlink_character_from_fact(
    state: State<'_, AppState>,
    fact_id: String,
    bible_entry_id: String,
) -> Result<(), String> {
    let db = state.get_db()?;
    let db = db.as_ref().ok_or("No project open")?;
    db.unlink_character_from_fact(&fact_id, &bible_entry_id)
}

#[tauri::command]
pub fn get_fact_characters(
    state: State<'_, AppState>,
    fact_id: String,
) -> Result<Vec<FactCharacter>, String> {
    let db = state.get_db()?;
    let db = db.as_ref().ok_or("No project open")?;
    db.get_fact_characters(&fact_id)
}

#[tauri::command]
pub fn get_character_knowledge_at_scene(
    state: State<'_, AppState>,
    bible_entry_id: String,
    scene_id: String,
) -> Result<Vec<Fact>, String> {
    let db = state.get_db()?;
    let db = db.as_ref().ok_or("No project open")?;
    db.get_character_knowledge_at_scene(&bible_entry_id, &scene_id)
}
